### ISSUE TITLE
Hardcoded units in another file?

### ISSUE THREAD

**Issue by kerimoyle** [cellml/libcellml#324]
I'd like to avoid duplicating hardcoded lists/dictionaries if possible.  Currently there are string vectors in the validator in functions:
- `isStandardPrefixName`
- `isStandardUnitName`
- `unitsAreEquivalent`

I'd like to either move them to another file or be able to share them, but they aren't enums so the enumerations.h file didn't seem like the right place.  Any suggestions?

```
Inside isStandardUnitName(const std::string &name)
    std::vector<std::string> standardUnitNames =  {
        "ampere", "becquerel", "candela", "celsius", "coulomb", "dimensionless", "farad", "gram", "gray",
        "henry", "hertz", "joule", "katal", "kelvin", "kilogram", "liter", "litre", "lumen", "lux",
        "meter", "metre", "mole", "newton", "ohm", "pascal", "radian", "second", "siemens", "sievert",
        "steradian", "tesla", "volt", "watt", "weber"
    };
   ... etc ... 

Inside isStandardPrefixName(const std::string &name)
    std::vector<std::string> prefixNames =
    {
        "atto", "centi", "deca", "deci", "exa", "femto", "giga", "hecto", "kilo", "mega", "micro", "milli",
        "nano", "peta", "pico", "tera", "yocto", "yotta", "zepto", "zetta"
    };
... etc ...

Inside unitsAreEquivalent()

standardList["farad"]["metre"] = -2.0;
standardList["farad"]["kilogram"] = -1.0;
standardList["farad"]["second"] = -4.0;
standardList["farad"]["ampere"] = 2.0;
    ... 
prefixList["yotta"] = 24;
prefixList["zetta"] = 21;
prefixList["exa"] = 18;
prefixList["peta"] = 15;
... etc ...
```

**Comment by agarny** [comment]
What about `utilities.[cpp|h]`?

**Comment by hsorby** [comment]
Or if you don't consider them to be utilities then `common.[cpp|h]`.  I think `utilities` is the best home though.

**Comment by kerimoyle** [comment]
Thanks both, I'll move to utilities :)

**Comment by kerimoyle** [comment]
A couple of things:
- do we want to include "celsius" in the list of standard units?  It (was) in the current `isStandardUnitName `function test but not in the specification list.  And if so, I'll make it equivalent to kelvin?
- I tried moving to the `utilities.h` file, but they generate linking errors there (even with adding preprocessor tests on top of the `#pragma once` ... just in case ...).  The errors disappear when they are in a separate `common.h` file though but I am stumped as to what the difference is!

Changes [here](https://github.com/kerimoyle/libcellml-keri/tree/validation_cleanup).

**Comment by MichaelClerx** [comment]
No to celsius, if we're not supporting offsets!

**Comment by kerimoyle** [comment]
Offsets don't make a difference to either the multiplication factor and the base units, so wouldn't appear in the equivalence tests anyway.  Overall am guessing it's not good form to include them, but they appear in several tests so it seemed that they had assumed to be "standard".  Happy to remove though :)

**Comment by agarny** [comment]
> * do we want to include "celsius" in the list of standard units?  It (was) in the current `isStandardUnitName `function test but not in the specification list.  And if so, I'll make it equivalent to kelvin?

`offset` is available in CellML 1.0/1.1, but **not** in CellML 2.0. So, `celsius` cannot and must not be be in the list of standard units (and should be removed from any test that includes it).

> * I tried moving to the `utilities.h` file, but they generate linking errors there (even with adding preprocessor tests on top of the `#pragma once` ... just in case ...).  The errors disappear when they are in a separate `common.h` file though but I am stumped as to what the difference is!

Impossible to tell what is wrong with so little information. On which platform are you getting those linking errors? What exactly are those linking errors?

> Changes [here](https://github.com/kerimoyle/libcellml-keri/tree/validation_cleanup).

Which ones exactly? You are pointing to your branch, not a specific commit as such.
